lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -0,0 +1,159 @@
+  std::vector<DeviceInfo> deviceInfo;
+  for (auto &device : devices_) {
+    DeviceInfo info = DeviceInfo();
+    info.availableMemory = device.second->getAvailableMemory();
+    deviceInfo.push_back(info);
+  }
+  // Optimize functions before passing to partitioner.
+  // Currently hardcoding inference.
+  if (backend_) {
+    for (auto F : M->getFunctions()) {
+      backend_->optimizeFunction(CompilationMode::Infer, F);
+    }
+  }
+  auto partitioner = Partitioner(M, deviceInfo);
+  // auto provisioner = llvm::make_unique<Provisioner>();
+  auto nodeList = std::move(partitioner.Partition());
+  auto result = provisioner_->provision(nodeList.roots, devices_, *M);
+
+  for (auto &node : nodeList.roots) {
+    roots_.emplace(node->name, std::move(node));
+  }
+  for (auto &node : nodeList.nodes) {
+    networks_.emplace(node->name, std::move(node));
+  }
+  return result;
+}
+
+void HostManager::removeNetwork(llvm::StringRef networkName) {
+  // Walk the tree for the given function, calling evict function for each node
+  // before removing it from networks_ which frees the node.
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  if (roots_.find(networkName) == roots_.end()) {
+    return;
+  }
+  std::queue<std::string> nodes;
+  std::set<std::string> allNodes;
+  for (auto &child : roots_[networkName]->children) {
+    nodes.push(child->name);
+    allNodes.insert(child->name);
+  }
+
+  while (!nodes.empty()) {
+    auto nodeName = nodes.front();
+    nodes.pop();
+    auto it = networks_.find(nodeName);
+    if (it != networks_.end()) {
+      for (auto &node : it->second->children) {
+        auto name = node->name;
+        if (allNodes.find(node->name) == allNodes.end()) {
+          nodes.push(node->name);
+          allNodes.insert(node->name);
+        }
+      }
+    }
+  }
+  roots_.erase(roots_.find(networkName));
+  for (auto &networkName : allNodes) {
+    auto networkIterator = networks_.find(networkName);
+    if (networkIterator != networks_.end()) {
+      devices_[networkIterator->second->deviceID]->evictNetwork(networkName);
+      networks_.erase(networkIterator);
+    }
+  }
+}
+
+bool HostManager::networkAdded(llvm::StringRef networkName) {
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  return roots_.find(networkName) != roots_.end();
+}
+
+void HostManager::clearHost() {
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  for (auto &it : devices_) {
+    it.second->stop();
+  }
+  for (auto &network : networks_) {
+    devices_[network.second->deviceID]->evictNetwork(network.second->name);
+  }
+  networks_.clear();
+  roots_.clear();
+
+  activeRequestCount_ = 0;
+}
+
+RunIdentifierTy HostManager::runNetwork(llvm::StringRef networkName,
+                                        std::unique_ptr<Context> context,
+                                        ResultCBTy callback) {
+
+  auto currentRun = totalRequestCount_++;
+  std::lock_guard<std::mutex> networkLock(networkLock_);
```

Comment:
does this lock need to be held while callback() is being called?

Reply:
In the callback we are good since activeRequestCount_ is atomic. It's needed here since we are accessing roots_.